File: RefraSin.TEPSolver.EquationSystem/Lagrangian.fs
module RefraSin.TEPSolver.EquationSystem.Lagrangian

open RefraSin.ParticleModel
open RefraSin.ProcessModel
open RefraSin.TEPSolver
open RefraSin.TEPSolver.ParticleModel
open RefraSin.TEPSolver.StepVectors
open System.Linq

let StateVelocityDerivativeNormal
    (conditions: IProcessConditions)
    (currentEstimation: StepVector)
    (node: NodeBase)
    : float =

    let gibbsTerm = -node.GibbsEnergyGradient.Normal * (1.0 + currentEstimation.Lambda1)

    let requiredConstraintsTerm =
        node.VolumeGradient.Normal * currentEstimation[node].LambdaVolume

    let contactTerm =
        match node with
        | :? ContactNodeBase as cn ->
            cn.ContactDistanceGradient.Normal * currentEstimation[cn].LambdaContactDistance
            + cn.ContactDirectionGradient.Normal
              * currentEstimation[cn].LambdaContactDirection
        | _ -> 0.0

    gibbsTerm + contactTerm + requiredConstraintsTerm

let StateVelocityDerivativeTangential
    (conditions: IProcessConditions)
    (currentEstimation: StepVector)
    (node: ContactNodeBase)
    : float =

    let gibbsTerm = -node.GibbsEnergyGradient.Normal * (1.0 + currentEstimation.Lambda1)

    let requiredConstraintsTerm =
        node.VolumeGradient.Normal * currentEstimation[node].LambdaVolume

    let contactTerm =
        node.ContactDistanceGradient.Normal
        * currentEstimation[node].LambdaContactDistance
        + node.ContactDirectionGradient.Normal
          * currentEstimation[node].LambdaContactDirection

    gibbsTerm + contactTerm + requiredConstraintsTerm

let FluxDerivative (conditions: IProcessConditions) (currentEstimation: StepVector) (node: NodeBase) : float =
    let dissipationTerm =
        2.0 * conditions.GasConstant * conditions.Temperature
        / (node.Particle.Material.MolarVolume
           * node.Particle.Material.EquilibriumVacancyConcentration)
        * node.SurfaceDistance.ToUpper
        * currentEstimation[node].FluxToUpper
        / node.SurfaceDiffusionCoefficient.ToUpper
        * currentEstimation.Lambda1

    let thisRequiredConstraintsTerm = currentEstimation[node].LambdaVolume
    let upperRequiredConstraintsTerm = currentEstimation[node.Upper].LambdaVolume

    -dissipationTerm - thisRequiredConstraintsTerm + upperRequiredConstraintsTerm

let RequiredConstraint (conditions: IProcessConditions) (currentEstimation: StepVector) (node: NodeBase) : float =
    let normalVolumeTerm = node.VolumeGradient.Normal * currentEstimation[node].NormalDisplacement

    let tangentialVolumeTerm =
        match node with
        | :? ContactNodeBase as cn -> node.VolumeGradient.Tangential * currentEstimation[cn].TangentialDisplacement
        | _ -> 0.0

    let fluxTerm = currentEstimation[node].FluxToUpper - currentEstimation[node.Lower].FluxToUpper

    normalVolumeTerm + tangentialVolumeTerm - fluxTerm

let ContactNodeEquations
    (conditions: IProcessConditions)
    (currentEstimation: StepVector)
    (contact: ParticleContact)
    (involvedNodes: ContactNodeBase seq)
    : float seq =
    seq {
        for n in involvedNodes do
            yield StateVelocityDerivativeTangential conditions currentEstimation n
            yield StateVelocityDerivativeTangential conditions currentEstimation n.ContactedNode

            yield! ContactConstraints conditions currentEstimation contact n

            yield
                currentEstimation[n].LambdaContactDistance
                - currentEstimation[n.ContactedNode].LambdaContactDistance

            yield
                currentEstimation[n].LambdaContactDirection
                - currentEstimation[n.ContactedNode].LambdaContactDirection
    }

let NodeEquations
    (conditions: IProcessConditions)
    (currentState: SolutionState)
    (currentEstimation: StepVector)
    : float seq =

    seq {
        for n in currentState.Nodes do
            yield StateVelocityDerivativeNormal conditions currentEstimation n
            yield FluxDerivative conditions currentEstimation n
            yield RequiredConstraint conditions currentEstimation n
    }

let ContactEquations
    (conditions: IProcessConditions)
    (currentState: SolutionState)
    (currentEstimation: StepVector)
    : float seq =

    seq {
        for c in currentState.Contacts do
            let involvedNodes =
                query {
                    for n in c.From.Nodes.OfType<ContactNodeBase>() do
                        where (n.ContactedParticleId = c.To.Id)
                        select n
                }

            yield! ContactNodeEquations conditions currentEstimation c involvedNodes

            yield ContactDistanceDerivative conditions currentEstimation c involvedNodes
            yield ContactDistanceDerivative conditions currentEstimation c involvedNodes
            yield ContactDistanceDerivative conditions currentEstimation c involvedNodes
    }

let DissipationEquality
    (conditions: IProcessConditions)
    (currentState: SolutionState)
    (currentEstimation: StepVector)
    : float =

    let dissipationNormal =
        query {
            for n in currentState.Nodes do
                sumBy (-n.GibbsEnergyGradient.Normal * currentEstimation[n].NormalDisplacement)
        }

    let dissipationTangential =
        query {
            for n in currentState.Nodes.OfType<ContactNodeBase>() do
                sumBy (-n.GibbsEnergyGradient.Tangential * currentEstimation[n].TangentialDisplacement)
        }

    let dissipationFunction =
        query {
            for p in currentState.Particles do
                sumBy (
                    query {
                        for n in p.Nodes do
                            sumBy (
                                n.SurfaceDistance.ToUpper * currentEstimation[n].FluxToUpper ** 2
                                / n.SurfaceDiffusionCoefficient.ToUpper
                            )
                    }
                    / (p.Material.MolarVolume * p.Material.EquilibriumVacancyConcentration)
                )
        }
        * conditions.GasConstant
        * conditions.Temperature

    dissipationNormal + dissipationTangential - dissipationFunction

let YieldEquations
    (conditions: IProcessConditions)
    (currentState: SolutionState)
    (currentEstimation: StepVector)
    : float seq =
    seq {
        yield! NodeEquations conditions currentState currentEstimation
        yield! ContactEquations conditions currentState currentEstimation
        yield DissipationEquality conditions currentState currentEstimation
    }

File: RefraSin.TEPSolver/EquationSystem/Lagrangian.cs
using MathNet.Numerics.LinearAlgebra;
using MathNet.Numerics.LinearAlgebra.Double;
using RefraSin.ParticleModel;
using RefraSin.TEPSolver.ParticleModel;
using RefraSin.TEPSolver.StepVectors;
using static System.Math;
using static RefraSin.TEPSolver.EquationSystem.Helper;
using NeckNode = RefraSin.TEPSolver.ParticleModel.NeckNode;
using Particle = RefraSin.TEPSolver.ParticleModel.Particle;
using ParticleContact = RefraSin.TEPSolver.ParticleModel.ParticleContact;

namespace RefraSin.TEPSolver.EquationSystem;

public static class Lagrangian
{
    public static Vector<double> EvaluateAt(SolutionState currentState, StepVector stepVector)
    {
        var evaluation = YieldEquations(currentState, stepVector).ToArray();

        if (evaluation.Any(x => !double.IsFinite(x)))
        {
            throw new InvalidOperationException(
                "One ore more components of the gradient evaluated to an infinite value."
            );
        }

        return new DenseVector(evaluation);
    }

    public static IEnumerable<double> YieldEquations(
        SolutionState currentState,
        StepVector stepVector
    ) =>
        Join(
            currentState.Particles.SelectMany(p => YieldParticleBlockEquations(p, stepVector)),
            YieldBorderBlockEquations(currentState, stepVector)
        );

    public static IEnumerable<double> YieldParticleBlockEquations(
        Particle particle,
        StepVector stepVector
    ) => YieldNodeEquations(particle.Nodes, stepVector);

    public static IEnumerable<double> YieldBorderBlockEquations(
        SolutionState currentState,
        StepVector stepVector
    ) => YieldLinearBorderBlockEquations(currentState, stepVector).Append(DissipationEquality(currentState, stepVector));
    
    public static IEnumerable<double> YieldLinearBorderBlockEquations(
        SolutionState currentState,
        StepVector stepVector
    ) => YieldContactsEquations(currentState.ParticleContacts, stepVector);

    public static IEnumerable<double> YieldNodeEquations(
        IEnumerable<NodeBase> nodes,
        StepVector stepVector
    )
    {
        foreach (var node in nodes)
        {
            if (node is not ContactNodeBase)
            {
                yield return StateVelocityDerivativeNormal(stepVector, node);
                yield return FluxDerivative(stepVector, node);
                yield return RequiredConstraint(stepVector, node);
            }
        }
    }

    public static IEnumerable<double> YieldContactsEquations(
        IEnumerable<ParticleContact> contacts,
        StepVector stepVector
    ) =>
        contacts.SelectMany(contact =>
            Join(
                YieldContactNodesEquations(stepVector, contact),
                YieldContactAuxiliaryDerivatives(stepVector, contact)
            )
        );

    public static IEnumerable<double> YieldContactNodesEquations(
        StepVector stepVector,
        ParticleContact contact
    )
    {
        foreach (var contactNode in contact.FromNodes)
        {
            yield return ContactConstraintDistance(stepVector, contact, contactNode);
            yield return ContactConstraintDirection(stepVector, contact, contactNode);

            yield return StateVelocityDerivativeNormal(stepVector, contactNode);
            yield return FluxDerivative(stepVector, contactNode);
            yield return RequiredConstraint(stepVector, contactNode);
            
            yield return StateVelocityDerivativeNormal(stepVector, contactNode.ContactedNode);
            yield return FluxDerivative(stepVector, contactNode.ContactedNode);
            yield return RequiredConstraint(stepVector, contactNode.ContactedNode);

            if (contactNode is NeckNode neckNode)
            {
                yield return StateVelocityDerivativeTangential(stepVector, neckNode);
                yield return StateVelocityDerivativeTangential(stepVector, neckNode.ContactedNode);
            }
        }
    }

    public static IEnumerable<double> YieldContactAuxiliaryDerivatives(
        StepVector stepVector,
        ParticleContact contact
    )
    {
        yield return ParticleRadialDisplacementDerivative(stepVector, contact);
        yield return ParticleAngleDisplacementDerivative(stepVector, contact);
    }

    public static double ParticleRadialDisplacementDerivative(
        StepVector stepVector,
        ParticleContact contact
    ) => contact.FromNodes.Sum(stepVector.LambdaContactDistance);

    public static double ParticleAngleDisplacementDerivative(
        StepVector stepVector,
        ParticleContact contact
    ) => contact.FromNodes.Sum(stepVector.LambdaContactDirection);

    public static double StateVelocityDerivativeNormal(StepVector stepVector, NodeBase node)
    {
        var gibbsTerm =
            node.GibbsEnergyGradient.Normal * (1 + stepVector.LambdaDissipation());
        var requiredConstraintsTerm = node.VolumeGradient.Normal * stepVector.LambdaVolume(node);

        double contactTerm = 0;

        if (node is ContactNodeBase contactNode)
        {
            contactTerm =
                contactNode.ContactDistanceGradient.Normal
              * stepVector.LambdaContactDistance(contactNode)
              + contactNode.ContactDirectionGradient.Normal
              * stepVector.LambdaContactDirection(contactNode);
        }

        return -gibbsTerm + requiredConstraintsTerm - contactTerm;
    }

    public static double StateVelocityDerivativeTangential(StepVector stepVector, NeckNode node)
    {
        var gibbsTerm =
            node.GibbsEnergyGradient.Tangential * (1 + stepVector.LambdaDissipation());
        var requiredConstraintsTerm =
            node.VolumeGradient.Tangential * stepVector.LambdaVolume(node);
        var contactTerm =
            node.ContactDistanceGradient.Tangential * stepVector.LambdaContactDistance(node)
          + node.ContactDirectionGradient.Tangential * stepVector.LambdaContactDirection(node);

        return -gibbsTerm + requiredConstraintsTerm - contactTerm;
    }

    public static double FluxDerivative(StepVector stepVector, NodeBase node)
    {
        var dissipationTerm =
            2
          * node.Particle.VacancyVolumeEnergy
          * node.SurfaceDistance.ToUpper
          / node.SurfaceDiffusionCoefficient.ToUpper
          * stepVector.FluxToUpper(node)
          * stepVector.LambdaDissipation();
        var thisRequiredConstraintsTerm = stepVector.LambdaVolume(node);
        var upperRequiredConstraintsTerm = stepVector.LambdaVolume(node.Upper);

        return -dissipationTerm - thisRequiredConstraintsTerm + upperRequiredConstraintsTerm;
    }

    public static double RequiredConstraint(StepVector stepVector, NodeBase node)
    {
        var normalVolumeTerm = node.VolumeGradient.Normal * stepVector.NormalDisplacement(node);
        var tangentialVolumeTerm = 0.0;

        if (node is NeckNode neckNode)
        {
            tangentialVolumeTerm =
                node.VolumeGradient.Tangential * stepVector.TangentialDisplacement(neckNode);
        }

        var fluxTerm = stepVector.FluxToUpper(node) - stepVector.FluxToUpper(node.Lower);

        return normalVolumeTerm + tangentialVolumeTerm - fluxTerm;
    }

    public static double DissipationEquality(SolutionState currentState, StepVector stepVector)
    {
        var dissipationNormal = currentState
            .Nodes.Select(n => -n.GibbsEnergyGradient.Normal * stepVector.NormalDisplacement(n))
            .Sum();

        var dissipationTangential = currentState
            .Nodes.OfType<NeckNode>()
            .Select(n => -n.GibbsEnergyGradient.Tangential * stepVector.TangentialDisplacement(n))
            .Sum();

        var dissipationFunction = currentState
            .Nodes.Select(n =>
                n.Particle.VacancyVolumeEnergy
              * n.SurfaceDistance.ToUpper
              * Pow(stepVector.FluxToUpper(n), 2)
              / n.SurfaceDiffusionCoefficient.ToUpper
            )
            .Sum();

        return dissipationNormal + dissipationTangential - dissipationFunction;
    }

    public static double ContactConstraintDistance(
        StepVector stepVector,
        IParticleContact contact,
        ContactNodeBase node
    ) =>
        stepVector.RadialDisplacement(contact)
      - node.ContactDistanceGradient.Normal * stepVector.NormalDisplacement(node)
      - node.ContactedNode.ContactDistanceGradient.Normal
      * stepVector.NormalDisplacement(node.ContactedNode)
      - (
            node is NeckNode
                ? node.ContactDistanceGradient.Tangential * stepVector.TangentialDisplacement(node)
                + node.ContactedNode.ContactDistanceGradient.Tangential
                * stepVector.TangentialDisplacement(node.ContactedNode)
                : 0
        );

    public static double ContactConstraintDirection(
        StepVector stepVector,
        IParticleContact contact,
        ContactNodeBase node
    ) =>
        stepVector.AngleDisplacement(contact)
      - node.ContactDirectionGradient.Normal * stepVector.NormalDisplacement(node)
      - node.ContactedNode.ContactDirectionGradient.Normal
      * stepVector.NormalDisplacement(node.ContactedNode)
      - (
            node is NeckNode
                ? node.ContactDirectionGradient.Tangential * stepVector.TangentialDisplacement(node)
                + node.ContactedNode.ContactDirectionGradient.Tangential
                * stepVector.TangentialDisplacement(node.ContactedNode)
                : 0
        );
}